lib/ai/model/agent.go

```diff
@@ -357,45 +352,61 @@ type PlanOutput struct {
 
 // parsePlanningOutput parses the output of the model after asking it to plan its next action
 // and returns the appropriate event type or an error.
-func parsePlanningOutput(deltas <-chan string) (*AgentAction, *agentFinish, error) {
+func parsePlanningOutput(deltas <-chan string) (*AgentAction, *agentFinish, TokenCounter, error) {
 	var text string
 	for delta := range deltas {
 		text += delta
 
 		if strings.HasPrefix(text, finalResponseHeader) {
 			parts := make(chan string)
+			streamingTokenCounter := NewAsynchronousTokenCounter(text)
 			go func() {
 				defer close(parts)
+				defer func() {
+					errCount := streamingTokenCounter.Finish()
```

Comment:
I don't think I fully understand this pattern. You create the counter, you add all tokens and then you call `TokenCount()` in one thread and `Finish()` in the other?  Why `TokenCount()` cannot just count all tokens and return the value?

Reply:
I revamped the logic to remove the wait logic and the Finished() function. I thought we were streaming completely asynchronously to the front end, but [we are waiting for the stream to end in `assist.go`](https://github.com/gravitational/teleport/blob/74067b8f38ba7c27690459115c5514048020edac/lib/assist/assist.go#L342). Any `TokenCount()` invocation after this line will return the correct count.